### PR TITLE
Modernize Python 2 code to prepare for Python 3

### PR DESCRIPTION
**Pull request by cclauss**
$ [__futurize --stage1 -w .__](https://python-future.org/automatic_conversion.html)

I do not understand how to deliver 100% test coverage while adhering to the Python porting best practice [__use feature detection instead of version detection__](https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection).

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@ botocore==1.10.5
 dnspython==1.15.0
 docutils==0.14
 dyn==1.8.1
-futures==3.2.0
+futures==3.2.0; python_version < '3.0'
```
[the remark this answers is not in the document]
Comment (cclauss) [reply requirements.txt]: [__pip 1.4__](https://hynek.me/articles/conditional-python-dependencies) current is __pip 19.1.1__.

File: octodns/provider/route53.py [hunk 2]
```diff
@@ -14,10 +14,17 @@
 import logging
 import re
 
+from six import text_type
+
 from ..record import Record, Update
 from ..record.geo import GeoCodes
 from .base import BaseProvider
 
+try:
```
[the remark this answers is not in the document]
Comment (cclauss) [reply octodns/provider/route53.py]: Done.

File: octodns/manager.py [hunk 3]
```diff
@@ -276,9 +276,9 @@ def sync(self, eligible_zones=[], eligible_targets=[], dry_run=True,
 
             try:
                 sources = [self.providers[source] for source in sources]
-            except KeyError:
+            except KeyError as e:
                 raise Exception('Zone {}, unknown source: {}'.format(zone_name,
-                                                                     source))
+                                                                     e))
```
[the remark this answers is not in the document]
Comment (cclauss) [reply octodns/manager.py]: Done.  I like this approach much better than the comprehension.

**Comment by cclauss** [comment]
See the original commit message?  How do we get to 100% test coverage (or lower that threshold)?

File: .travis.yml [hunk 4]
```diff
@@ -1,6 +1,12 @@
 language: python
-python:
-  - 2.7
+matrix:
+  include:
+    - python: 2.7
+    - python: 3.7
+      dist: xenial  # required for Python >= 3.7 on Travis CI
+  allow_failures:
+    - python: 3.7
```
[the remark this answers is not in the document]
Comment (cclauss) [reply .travis.yml]: I usually leave __allow_failures__ in place until the tests pass.

**Comment by ross** [comment]
> See the original commit message... How do we get to 100% test coverage (or lower that threshold)?

Yeah, there's be no way to get both python2 and python3 happy with that block so disabling coverage makes sense:

```python
# TODO: remove when Python 2.x is no longer supported
try:  # pragma: no cover
    cmp
except NameError:  # pragma: no cover
    def cmp(x, y):
        return (x > y) - (x < y)
```

**Comment by cclauss** [comment]
On Python 3:
```
Ran 90 tests in 0.943s
FAILED (errors=32, failures=11)
The command "./script/cibuild" exited with 1.
```

__allow_failures__ has its place.  I encourage you to merge this PR as is and then we can work on the 32 + 11 in other PRs.

**Comment by ross** [comment]
Pushed up some passes at moving this further along over in https://github.com/github/octodns/pull/384. It's still not complete, but a lot closer. Main hiccup is dependent libraries that aren't python3 compat, the one I'm sure about is `incf.countryutils`. Azure may be happier with newer versions, but I haven't had a chance to dig into that yet.

Anyway opened https://github.com/github/octodns/pull/384 rather than push substantial changes here so it probably make sense to consider ^ as a continuation of this PR. Going to close this one and follow up discussion can happen there. Feel free to open PRs against that PR if you end up having further commits.

**Pull request closed without merging**